backend/groth16/bn254/marshal_gpu.go

```diff
@@ -0,0 +1,381 @@
+func (pk *ProvingKey) WriteRawTo(w io.Writer) (n int64, err error) {
+	return pk.writeTo(w, true)
+}
+
+func (pk *ProvingKey) writeTo(w io.Writer, raw bool) (int64, error) {
+	n, err := pk.Domain.WriteTo(w)
+	if err != nil {
+		return n, err
+	}
+
+	var enc *curve.Encoder
+	if raw {
+		enc = curve.NewEncoder(w, curve.RawEncoding())
+	} else {
+		enc = curve.NewEncoder(w)
+	}
+	nbWires := uint64(len(pk.InfinityA))
+
+	toEncode := []interface{}{
+		&pk.G1.Alpha,
+		&pk.G1.Beta,
+		&pk.G1.Delta,
+		pk.G1.A,
+		pk.G1.B,
+		pk.G1.Z,
+		pk.G1.K,
+		&pk.G2.Beta,
+		&pk.G2.Delta,
+		pk.G2.B,
+		nbWires,
+		pk.NbInfinityA,
+		pk.NbInfinityB,
+		pk.InfinityA,
+		pk.InfinityB,
+		uint32(len(pk.CommitmentKeys)),
+	}
+
+	for _, v := range toEncode {
+		if err := enc.Encode(v); err != nil {
+			return n + enc.BytesWritten(), err
+		}
+	}
+
+	for i := range pk.CommitmentKeys {
+		var (
+			n2  int64
+			err error
+		)
+		if raw {
+			n2, err = pk.CommitmentKeys[i].WriteRawTo(w)
+		} else {
+			n2, err = pk.CommitmentKeys[i].WriteTo(w)
+		}
+
+		n += n2
+		if err != nil {
+			return n, err
+		}
+	}
+
+	return n + enc.BytesWritten(), nil
+
+}
+
+// ReadFrom attempts to decode a ProvingKey from reader
+// ProvingKey must be encoded through WriteTo (compressed) or WriteRawTo (uncompressed)
+// note that we don't check that the points are on the curve or in the correct subgroup at this point
+func (pk *ProvingKey) ReadFrom(r io.Reader) (int64, error) {
+	return pk.readFrom(r)
+}
+
+// UnsafeReadFrom behaves like ReadFrom excepts it doesn't check if the decoded points are on the curve
+// or in the correct subgroup
+func (pk *ProvingKey) UnsafeReadFrom(r io.Reader) (int64, error) {
+	return pk.readFrom(r, curve.NoSubgroupChecks())
+}
+
+func (pk *ProvingKey) readFrom(r io.Reader, decOptions ...func(*curve.Decoder)) (int64, error) {
+	n, err := pk.Domain.ReadFrom(r)
+	if err != nil {
+		return n, err
+	}
+
+	dec := curve.NewDecoder(r, decOptions...)
+
+	var nbWires uint64
+	var nbCommitments uint32
+
+	toDecode := []interface{}{
+		&pk.G1.Alpha,
+		&pk.G1.Beta,
+		&pk.G1.Delta,
+		&pk.G1.A,
+		&pk.G1.B,
+		&pk.G1.Z,
+		&pk.G1.K,
+		&pk.G2.Beta,
+		&pk.G2.Delta,
+		&pk.G2.B,
+		&nbWires,
+		&pk.NbInfinityA,
+		&pk.NbInfinityB,
+	}
+
+	for _, v := range toDecode {
+		if err := dec.Decode(v); err != nil {
+			return n + dec.BytesRead(), err
+		}
+	}
+	pk.InfinityA = make([]bool, nbWires)
+	pk.InfinityB = make([]bool, nbWires)
+
+	if err := dec.Decode(&pk.InfinityA); err != nil {
+		return n + dec.BytesRead(), err
+	}
+	if err := dec.Decode(&pk.InfinityB); err != nil {
+		return n + dec.BytesRead(), err
+	}
+	if err := dec.Decode(&nbCommitments); err != nil {
+		return n + dec.BytesRead(), err
+	}
+
+	pk.CommitmentKeys = make([]pedersen.ProvingKey, nbCommitments)
+	for i := range pk.CommitmentKeys {
+		n2, err := pk.CommitmentKeys[i].ReadFrom(r)
+		n += n2
+		if err != nil {
+			return n, err
+		}
+	}
+
+	if err := pk.setupDevicePointers(); err != nil {
```

Comment:
>what happens for example when we unmarshal multiple provingkeys, would the GPU memory be overwritten when we later want to prove with the first key

As long as there is a reference to the original pk, it should contain the GPU memory pointers for that specific key. The second pk has its own GPU memory pointers for its data. 

As long as the GPU memory is large enough to hold all of the necessary data for multiple pks and as long as those pks still have CPU references, the GPU memory shouldn't be overwritten

>I understand if we would do it during setup-time, then we get duplicate copies

Why is this the case? Each pk has its own reference to its data in GPU memory. If we are loading/setting up the same pk multiple times then yes I can see duplications happening. Whats the use case for loading the same pk multiple times though?

>And I think it also skews the benchmark results a bit - I would include RAM-GPU memory copies part of the time cost, even if it can be amortized

🤔 For constant values that are computed once for a given circuit I don't see a reason why these times should be added to the prove timing; similar to the rest of setting up pk

>And by removing it from here we can unify marshal.go between GPU and non-GPU version.

That would be nice. We can move this to Prove and check pk for valid pointers to GPU; if they don't exist then transfer the data otherwise continue as we do now.